compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -559,44 +559,42 @@ public void visit(BLangTransaction transactionNode, AnalyzerData data) {
         //Check whether transaction statement occurred in a transactional scope
         if (data.transactionalFuncCheckStack.peek()) {
             this.dlog.error(transactionNode.pos,
-                            DiagnosticErrorCode.TRANSACTION_CANNOT_BE_USED_WITHIN_TRANSACTIONAL_SCOPE);
+                    DiagnosticErrorCode.TRANSACTION_CANNOT_BE_USED_WITHIN_TRANSACTIONAL_SCOPE);
             return;
         }
-
-        data.errorTypes.push(new LinkedHashSet<>());
-
         boolean previousWithinTxScope = data.withinTransactionScope;
         int previousCommitCount = data.commitCount;
         int previousRollbackCount = data.rollbackCount;
         boolean prevCommitRollbackAllowed = data.commitRollbackAllowed;
         data.commitRollbackAllowed = true;
         data.commitCount = 0;
         data.rollbackCount = 0;
-
         data.withinTransactionScope = true;
-
         data.loopWithinTransactionCheckStack.push(false);
         data.returnWithinTransactionCheckStack.push(false);
         data.transactionCount++;
+        boolean onFailExists = transactionNode.onFailClause != null;
         boolean failureHandled = data.failureHandled;
-        if (!failureHandled) {
-            data.failureHandled = transactionNode.onFailClause != null;
+        if (onFailExists) {
+            data.errorTypes.push(new LinkedHashSet<>());
+            data.failureHandled = true;
         }
         analyzeNode(transactionNode.transactionBody, data);
         data.failureHandled = failureHandled;
         if (data.commitCount < 1) {
             this.dlog.error(transactionNode.pos, DiagnosticErrorCode.INVALID_COMMIT_COUNT);
         }
-
         data.transactionCount--;
         data.withinTransactionScope = previousWithinTxScope;
         data.commitCount = previousCommitCount;
         data.rollbackCount = previousRollbackCount;
         data.commitRollbackAllowed = prevCommitRollbackAllowed;
         data.returnWithinTransactionCheckStack.pop();
         data.loopWithinTransactionCheckStack.pop();
-        analyzeOnFailClause(transactionNode.onFailClause, data);
-        data.errorTypes.pop();
+        if (onFailExists) {
+            transactionNode.transactionBody.failureBreakMode = getPossibleBreakMode(!data.errorTypes.peek().isEmpty());
+            analyzeOnFailClause(transactionNode.onFailClause, data);
```

Comment:
Shall we move this into a separate method?

Reply:
Updated in daeb71d752dd2b0ec1b555a64009727b164a51b2